### PR TITLE
⬆️ Electron 3 (Take Two)

### LINKED ISSUE #544: Freeze when dragging tabs (confirmed with 1.32.1)
### Prerequisites

* [X ] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

Atom completely freezes when dragging tab.

### Steps to Reproduce

1. Start atom with no options (or with `--safe`). (The problem goes away with `--disable-gpu`.)
2. Drag a pane.

**Expected behavior:**
Moved pane.

**Actual behavior:**
Atom hangs.  Sometimes it steals the window cursor, making it hard to kill.

**Reproduces how often:**
100%

### Versions
Atom version is 1.32.1, amd64, from atom.io.
OS is Fedora 29, with Gnome and Wayland.

### Additional Information
This appears to be Issue #541, confirmed with latest official release.

### LINKED ISSUE #6666: no mnemonics in menus (File, Edit, ...)
My system:
Linux
Desktop: KDE 4.14.6

In the menus there are any mnemonics shown at all!
(the menu buttons themselves show mnemonics when pressing ALT)
E.g. in File menu:

```
New Window
New File
Open File
...
```

don't show any mnemonics. Also when pressing "ALT" while menu is opened.
Nevertheless, mnemonics seem to be there (but not shown). If I press "n" while in File menu (found out by guessing) Atom opens a new file.

### LINKED ISSUE #18535: Items in the Menu Bar are invisible on Ubuntu 18.10 when atom window is focused.
### Prerequisites

* [ ] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: https://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

Items in the Menu Bar are invisible on Ubuntu 18.10 when atom window is focused.

### Steps to Reproduce

1. Launch Atom.

**Expected behavior:** Menu Bar items (File, Edit, etc) visible.

**Actual behavior:** Menu Bar items invisible.

**Reproduces how often:** Always.

### Versions
Atom    : 1.33.0
Electron: 2.0.11
Chrome  : 61.0.3163.100
Node    : 8.9.3

apm  2.1.2
npm  6.2.0
node 8.9.3 x64
atom 1.33.0
python 2.7.15+
git 2.19.1

Distributor ID:	Ubuntu
Description:	Ubuntu 18.10
Release:	18.10
Codename:	cosmic

Linux buddy 4.18.0-11-generic #12-Ubuntu SMP Tue Oct 23 19:22:37 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

Ubuntu is apt-upgraded as of today.

### Additional Information
screenshot from 2018-12-03 13-09-52
screenshot from 2018-12-03 13-11-23

### PR DESCRIPTION
**Pull request by jasonrudolph**
### Description of the Change

\#18603 and #18815 upgraded Atom from Electron 2 to Electron 3, but we unfortunately experienced frequent crashes following those changes. As a short-term solution, https://github.com/atom/atom/pull/18908 reverted those changes.

This pull request restores the changes from #18603 and #18815. (It does so by reverting the changes from #18908.) Before we merge this pull request, we'll need to resolve the crashes that led to us having to revert the previous Electron 3 upgrade.

### Issues to resolve

I _think_ these issues are all caused by the upgrade to Electron 3. If we discover that they're not directly related to the Electron 3 upgrade, we can remove them from this list and address them separately.

- [x] https://github.com/atom/atom/issues/18790
- [x] https://github.com/atom/atom/issues/18837
- [x] https://github.com/atom/atom/issues/18840
- [x] https://github.com/atom/atom/issues/18841
- [ ] https://github.com/atom/atom/issues/18915
- [x] https://github.com/atom/atom/issues/18882
- [x] https://github.com/atom/teletype/issues/438
- [x] Blurry font rendering (https://github.com/atom/atom/issues/19356)

Fixes #6666.
Fixes #18535.
Fixes atom/tabs#544

### CODE REVIEW

**Comment by Arcanemagus** [comment]
https://github.com/atom/atom/issues/18840 also showed up with the Electron upgrade builds.

**Comment by jasonrudolph** [comment]
> #18840 also showed up with the Electron upgrade builds.

Thanks, @Arcanemagus. I've added it to the list in the PR body.

**Comment by carlos157oliveira** [comment]
Hi, this is the first issue I've been assigned to triage by CodeTriage. This is still a issue? I see there is no recent activity. Would the maintainer consider closing this topic?

File: package.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
     "url": "https://github.com/atom/atom/issues"
   },
   "license": "MIT",
-  "electronVersion": "2.0.18",
+  "electronVersion": "3.1.8",
```
Comment (mfonville) [inline package.json]: @rafeca 
Electron 3.1.9 is also available with some small fixes: https://github.com/electron/electron/releases/tag/v3.1.9

Comment (jasonrudolph) [reply package.json]: @mfonville: Thanks! Updated in b2a4810eaf0d2f07acff056e03d6159030c0da0d.

**Comment by jasonrudolph** [comment]
Since Electron 3 uses Node 10, should this pull request update the build scripts to use Node 10?

For example, should we upgrade this build step to use Node 10?

https://github.com/atom/atom/blob/6e2c0b0732bc4a8ef1fdbfe40e30b553a0215dbb/script/vsts/platforms/macos.yml#L13-L17

**Comment by 50Wliu** [comment]
> For example, should we upgrade this build step to use Node 10?

Yes, I'm pretty sure they should. Electron 3 uses 10.2.0, though it might be best to go with [10.2.1](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V10.md#10.2.1) instead, as 10.2.0 has issues compiling native modules.

**Comment by rafeca** [comment]
I spoke with @daviwil earlier and he mentioned that upgrading the version of Node.js on CI is not necessary to do at the same time that upgrading to electron 3, since apm will take care of using the appropriate  node version through node-gyp.

We should do that as a followup task though, to be able to use the same language features between Atom and the build scripts

**Comment by mfonville** [comment]
Very happy to see that Electron 3.x is finally near. I hope the upgrade to later versions will go faster, because Electron only supports the last 3 releases and the release of Electron 6.x (and thus EOL of 3.x) is expected on July 30th: https://electronjs.org/blog/12-week-cadence

**Comment by 50Wliu** [comment]
Here's some weird shifting of content that I'm seeing after opening and closing the dev tools.
slight-content-shifting

**Comment by as-cii** [comment]
I also think we should consider looking into how text is being rendered in this new version before shipping this. Below you can find a zoomed-in screenshot of rendering the word `Lorem` in a variety of environments. Please, click on all of the images and quickly browse through the tabs to appreciate how stark the difference is for this pull request.

### Atom on Electron 3 (this pull request)
Atom on Electron 3 0

### VS Code on Electron 3 (latest published version)
VS Code

### Atom on Electron 2 (Atom 1.38.0-beta0)
Atom on Electron 2 0

You can reproduce the above tests by using Fira Code Retina, 15px. I suspect some of the styling we do in `TextEditorComponent` is interacting oddly with the new Chromium version.

**Comment by jasonrudolph** [comment]
> I also think we should consider looking into how text is being rendered in this new version before shipping it.

Thanks, @as-cii. I've added this to the task list in the PR body.

**Comment by jasonrudolph** [comment]
@50Wliu: It [looks like](https://user-images.githubusercontent.com/2988/57882070-d7ecfd80-77f0-11e9-8994-ef288bda4a60.png) you added this item to the task list in this PR body:

> The "restarting search from the top" overlay icon in find-and-replace doesn't seem to show up

I want to make sure we address that item, but I don't understand what it means exactly. (Sorry.) Can you add a link that provides more detail or a gif that demonstrates the issue?

**Comment by 50Wliu** [comment]
@jasonrudolph Never mind, I just tested it again and it's working :sweat_smile:. I forgot I needed to use it on a file that has a scrollbar.

**Comment by 50Wliu** [comment]
Okay, here's an actual regression. "Open link" (provided by the [link](https://github.com/atom/atom/tree/master/packages/link) package) no longer works.

Electron 3.1:
electron-3 1-broken-link

Electron 2.0 (Atom 1.37.0):
electron-2 0-working-link

**Comment by rafeca** [comment]
> Okay, here's an actual regression. "Open link" (provided by the link package) no longer works.

Thanks for reporting! I've created an issue for it: https://github.com/atom/atom/issues/19353

**Comment by rafeca** [comment]
> Okay, here's an actual regression. "Open link" (provided by the link package) no longer works.

I've done some investigation and this issue is not related to electron v3, since it's also present in current nightly versions of Atom which use electron v2 (more info in https://github.com/atom/atom/issues/19353), so I've removed it from the list of blocking issues for the upgrade.

**Comment by rafeca** [comment]
Since this issue is appearing also on v1.38-beta0, I've added it to the list of release blockers

**Comment by as-cii** [comment]
@rafeca and I think this pull request is ready to be merged. The sooner we do this, the more time we will have to catch bugs before it goes to Beta. Please, note that we are purposely ignoring https://github.com/atom/atom/issues/18915 for now because we haven't been able to reproduce it, and it's unclear how it would be related to the Electron upgrade.

Feedback on this is super welcome! Especially after this pull request gets merged, it would be useful to get as many 👀 on Nightly as possible to catch potential regressions that we didn't notice.

/cc: @atom/maintainers

**Comment by jasonrudolph** [comment]
> @rafeca and I think this pull request is ready to be merged.

Have you had a chance to run through the regression test plan? Once we've tested out all the items on that checklist, I'm 👍 on merging this pull request.

It's a pretty long checklist, so I'd be perfectly happy to divide and conquer with y'all.

**Comment by Arcanemagus** [comment]
Note: The `libgconf-2-4` dependency to be added in https://github.com/atom/atom/pull/19358 will need to be removed when this is merged, as Electron 3 doesn't require it.

**Comment by as-cii** [comment]
## Check for regressions experienced in previous upgrades

### Text Input/Keybindings

See how to [setup keyboard layouts](https://github.com/atom/maintainers-log/blob/master/how-tos/setup-keyboard-layouts.md).

- [ ] IME not working (https://github.com/atom/atom/issues/14911)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Install an IME layout, for example Pinyin simplified  
  1. Open Atom and use the layout you installed as the keyboard layout
  1. Example, if you picked Pinyin simplified type in `zhongwenshuru`

  **Expected:** in Pinyin simplified `中文输入` is expected

  gif

  **Issues we've seen:** Only shows latin first character and not every character typed.

  gif
  </details>

- [ ] macOS: Key binding with composition characters are ignored (https://github.com/atom/atom/issues/15189)
  <details>
  <summary>Repro steps</summary>

  1. Open your keymap.cson file and add the following:
      ```cson
      'atom-text-editor':
        # use hjkl; vim-like when pressing alt (word by word, paragraph by paragraph)
        # or alt (letter by letter)
        'alt-h': 'core:move-left'
        'alt-j': 'core:move-down'
        'alt-k': 'core:move-up'
        'alt-l': 'core:move-right'
      ```
  1. Save keymap.cson and reload atom.
  1. Try the new keymappings under `ABC - Extended` keyboard layout

  **Expected**: the keybindings work

  **Issues we've seen:** it types out the modified keys and ignores the key mapping
  </details>

- [ ] macOS: Composition characters mess up insertion point (https://github.com/atom/atom/issues/15344)
  <details>
  <summary>Repro steps</summary>

  1. Open keymap.cson and add the following:
      ```cson
      'atom-text-editor':
        'alt-n': 'core:move-down'
        'alt-o': 'core:move-down'
      ```
  1. Have the following in your buffer:
      ```
      Foo
      Bar
      ```
  1. Have the cursor at the beginning of file(before F character) Type alt-n, then arrow right, then insert a character. Observe that the character is inserted at the beginning of the line (where the alt-n left you).
  1. Repeat steps using <kbd>alt-o</kbd> instead of <kbd>alt-n</kbd>.

  **Expected**: Both alt-o and alt-n behave the same

  **Issues we've seen:**
  bug

- [ ] Ubuntu: Keystrokes involving <kbd>ctrl</kbd> resolve to the default layout instead of the active layout (https://github.com/atom/atom/issues/13170)

  <details>
  <summary>Repro steps</summary>

  **This should be tested on Linux with gnome**

  1. Install a Linux ditro with gnome, for example Ubuntu
  1. Install French AZERTY and English US QWERTY layouts
  1. Change the default layout to be French AZERTY
  1. Open Atom and switch to English US QWERTY layout
  1. Press ctrl-w

  **Expected**: `core:close` to dispatch (Or the keybinding-resolver to resolve to ctrl-w). Should be resolving to the keyboard layout that is chosen and not the OS default layout.

  **Issues we've seen:** `core:undo` dispatched because it was resolved as ctrl-z because AZERTY has Z where W is on QWERTY.
  </details>

- [ ] Other keyboard layouts on new Electron version
  - [ ] macOS: ABC - Extended
  - [ ] macOS: Slovak
  - [ ] macOS: U.S. International
  - [ ] macOS: Squirrel (https://github.com/rime/squirrel)
  - [ ] macOS: Full-Width Alphanumeric (Google) - Japanese layout  (https://www.google.co.jp/ime/)
  - [ ] Windows: U.S.
  - [ ] Linux: Pinyin Simplified

### UI

- [ ] tree-view drag image (https://github.com/atom/tree-view/pull/1054)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open a project in Atom with multiple files and a few folders
  1. Drag a file into a folder

  **Expected:**  
  image

  **Issues we've seen:**

  image
  </details>

- [ ] drag-and-drop indicator (https://github.com/atom/tree-view/pull/1055, https://github.com/atom/tabs/pull/426, https://github.com/atom/tabs/pull/437)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open Atom with a few folder projects on tree-view (you can do this on terminal in a directory that has existing folder and having a space between each folder. Example: `atom folderA folderB folderC`)
  1. Drag and drop all of the folders to reorder them on Atom

  **Expected:**
  The placeholder indicates where the folder/project is being dropped to.

  screen shot 2017-03-29 at 16 31 16

  **Issues we've seen:**
  No placeholder shows up after drag and dropping

  screen shot 2017-03-29 at 16 30 51
  </details>

- [ ] Large file rendering (https://github.com/atom/atom/issues/16591)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open a file that has over one million lines
  1. Scroll down to the bottom

  **Expected:** To not regress the number of lines that can be rendered. Rendering to be correct for lines past a certain point. Atom 1.25  can render around 800k to 1 million lines correctly.

  **Issues we've seen:** Increased number of lines rendered but bad rendering past a certain point

  bad renderer
  </details>

- [ ] Loss of subpixel AA when the cursor is at the end of long lines (https://github.com/atom/atom/issues/16889, https://github.com/atom/atom/pull/16595)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open [text-editor.js](https://github.com/atom/atom/blob/2e19cf4a7c0a4d624b1f9c1072cd292de112cdb0/src/text-editor.js) in Atom
  1. Copy the first 21 lines to a new untitled window
  1. Turn on soft wrapping and change language mode to javascript
  1. Resize the Window so it's soft wrapped
  1. On line 15 type `a` until the you reach the end of the window

  **Expected:** To not lose subpixel AA

  **Issues we've seen:** Loss of subpixel AA. Both when soft wrapping was [enabled](https://github.com/atom/atom/pull/16943) and [disabled](https://github.com/atom/atom/pull/16595).

  subpixel aa loss softwrap
  </details>

- [ ] Scrolling horizontally shift + scroll wheel (https://github.com/atom/atom/pull/12696#issuecomment-288615912)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open a file that has long lines
  1. Turn off soft wrapping and resize the window so you have a horizontal scrollbar
  1. Press shift and scroll in both directions using the mouse wheel

  **Expected:** File to scroll horizontally in both directions

  **Issues we've seen:** File does not scroll horizontally at all in any direction
  </details>

- [ ] Scrollbars misbehaving on the first file that is opened (https://github.com/atom/atom/pull/12696#issuecomment-251299649)
  <details>
  <summary>Repro steps</summary>

  1. Open Atom
  1. Open some file that has content enough to have a vertical scrollbar
  1. Close Atom
  1. Open Atom and look at the vertical scrollbar

  **Expected:** Scrollbar to be visible

  **Issues we've seen:** Scrollbar is not visible and is flickering when you are editing
  </details>

- [ ] Middle clicking on unsaved tab (https://github.com/atom/atom/issues/15197)
	- [ ] macOS
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open Atom and open a new unsaved file.
  1. Make edits to the unsaved file and middle click on the tab.

  **Expected:**
  Clicking save/cancel or the options on the dialog works.

  **Issues we've seen:**
  The UI and the dialog is unable to receive mouse clicks. You can still choose options via Keyboard, but not mouse.
  </details>

- [ ] Linux: Atom scrolls even when not focused (https://github.com/atom/atom/issues/15482)
  <details>
  <summary>Repro steps</summary>

  1. Open Atom with a large enough file that has a scrollbar
  1. Open a different application and place it over the Atom window
  1. Scroll with the mouse inside the other window
  1. Focus the Atom window

  **Expected:** Atom window should keep the original scroll position

  **Issues we've seen:** The Atom window scrolls after it is focused
  </details>

### Other

- [ ] Deprecation warnings (https://github.com/atom/atom/pull/12696#issuecomment-288871751)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  **This should be tested on macOS, Linux and Windows with community packages installed**

  1. Open Atom
  1. Open the Developer tools console on the master branch and check for deprecation warnings
  1. Open the Developer tools console on the electron upgrade branch and check for deprecation warnings
  1. Compare the two outputs

  **Expected:** No new deprecation warnings

  **Issues we've seen:** New deprecation warnings both from core and community packages
  </details>

- [ ] Supported Versions of OS (https://github.com/atom/atom/issues/15297)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Check the list of supported platforms [Electron Master](https://github.com/electron/electron/blob/master/docs/tutorial/supported-platforms.md)
  1. Check what version is listed as minimum for each platform on https://atom.io and https://atom.io/faq

  **Expected:** https://atom.io and https://atom.io/faq to display the same versions as the Electron documentation

  **Issues we've seen:** macOS version support changed and https://atom.io and https://atom.io/faq was not updated
  </details>

- [ ] macOS : Native tabs (https://github.com/atom/atom/pull/14711)
  <details>
  <summary>Repro steps</summary>

  1. Open two atom windows on mac
  1. Go to View in menus and check `Show Tab Bar`
  screen shot 2017-06-03 at 5 10 32 pm
  1. Do this for both Atom windows.
  1. Drag one atom window to the other via tab bar

  **Expected:**
  Native Tab bar shows up and able to drag different windows to each other.  Also make sure UI isn't messed up.

  **Issues we've seen:**
  Native Tab Bar doesn't work and messes up UI when enabled
  </details>

- [ ] Ubuntu with KDE: Menu uses 100% CPU (https://github.com/electron/electron/issues/8455, https://github.com/atom/atom/issues/13885)
  <details>
  <summary>Repro steps</summary>

  1. Install Atom on Ubuntu KDE. [[instructions](https://linuxconfig.org/how-to-install-kde-plasma-desktop-on-ubuntu-18-04-bionic-beaver-linux)]
  1. Enable KDE plasma global menu: Right-click on the desktop -> Add Panel => Application Menu Bar. [[demo](https://youtu.be/qo5fhiXAUbI)]
  1. Start Atom.
  1. Try to use the desktop menu.

  **Expected:**
  Desktop menu items work

  **Issues we've seen:**
  Desktop menu items do nothing when you click on things.
  </details>

- [ ] macOS: Slovak QWERTZ (https://github.com/atom/atom-keymap/issues/223)
  <details>
  <summary>Repro steps</summary>

  1. Install Slovak keyboard layout on mac
  1. Open keyboard viewer to see the keys are are suppose to show up and open keyboard resolver on atom.
  1. hit `cmd+'`.

  **Expected:**  Resolves to `cmd+'` like it shows on keyboard viewer on mac.

  **Issues we've seen:** It resolves to `ctrl+§`
  </details>

- [ ] macOS: IME jump (https://github.com/atom/atom/issues/15696)
  <details>
  <summary>Repro steps</summary>

  1. Install an IME layout, for example Chinese pinyin
  1. Type one character

  **Expected:**
  Expected the IME to be positioned at under the character you entered
  **Issues we've seen:** The IME window is in the top left corner. When you enter the second character it jumps to be positioned under the text
  </details>

**Comment by rafeca** [comment]
I've done the sanity checks on Linux and haven't found any regression related to this upgrade. I have some notes though about some items:

>  Large file rendering (#16591)

I've tried with a several million lines file and the editor becomes almost unresponsive (good thing is that Atom shows a warning saying that the file is too big and may cause issues with the editor). When I tried to scroll to ~800k line Atom ended up crashing.

This same behaviour happened on current Atom beta, so it's not an electron issue.

>  Scrolling horizontally shift + scroll wheel (#12696 (comment))

I'm not able to test this one because I don't have a wheel mouse.

>  Middle clicking on unsaved tab (#15197)

I'm not able to test this one because I don't have a wheel mouse.

**Comment by nathansobo** [comment]
Probable regression: https://github.com/atom/atom/issues/19372

**Comment by nathansobo** [comment]
Hate to have to say this, but I reverted this in 3da827b8b06128b2fbe94d5ee16d762871a7aa31 due to crashes related to #19372. The saga continues.

**Pull request merged**